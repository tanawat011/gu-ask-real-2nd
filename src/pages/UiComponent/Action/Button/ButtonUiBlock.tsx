import type { UiComponentDefaultProps } from 'types'

import React from 'react'

import tw from 'twin.macro'

import { Button } from 'components/Button'
import { ComponentDisplay, Title } from 'components/PageUiComponent'
import { PreCode } from 'components/PreCode'

const TwContainer = tw.div`mb-8`

export const ButtonUiBlock: React.FC<UiComponentDefaultProps> = (props) => {
  return (
    <TwContainer id={props.id}>
      <Title
        title={props.id}
        description={
          <>
            Button can stretch to fill the width by add <PreCode text='block' />{' '}
            prop.
          </>
        }
      />
      <ComponentDisplay
        components={
          <>
            <Button label='Non Block Button' />
            <Button label='Blocked Button' block />
            <Button label='Blocked Button' block disabled />
            <Button label='Blocked Button' block loading='Loading...' />
          </>
        }
        code={`import React from "react"

export const Component = () => {
  return (
    <>
    <Button label='Non Block Button' />
    <Button label='Blocked Button' block />
    <Button label='Blocked Button' block disabled />
    <Button label='Blocked Button' block loading='Loading...' />
    </>
  )
}`}
      />
    </TwContainer>
  )
}
