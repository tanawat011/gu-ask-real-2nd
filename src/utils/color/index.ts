export * from './hex2Rgb'
export * from './rgb2Hex'
