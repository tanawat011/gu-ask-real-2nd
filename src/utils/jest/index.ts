export * from './renderWithProviders'
export * from './twColor'
