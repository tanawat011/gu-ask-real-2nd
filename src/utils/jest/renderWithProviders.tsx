import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
