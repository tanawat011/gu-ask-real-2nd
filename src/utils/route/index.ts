export * from './menuGenerator'
export * from './renderRoutes'
export * from './routeGenerator'
