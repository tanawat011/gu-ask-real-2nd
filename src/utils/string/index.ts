export * from './camelToKebab'
export * from './camelToSentence'
export * from './replaceAll'
