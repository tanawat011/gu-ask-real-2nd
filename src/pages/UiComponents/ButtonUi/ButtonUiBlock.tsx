import type { LegacyRef } from 'react'
import { useEffect } from 'react'

import { useRecoilValue } from 'recoil'
import tw from 'twin.macro'

import { Button } from 'components/Button'
import { ComponentDisplay, Title } from 'components/PageUiComponent'
import { PreCode } from 'components/PreCode'
import { useDimensionChange } from 'hooks/useDimensionChange'
import { useElementDimension } from 'hooks/useElementDimension'
import { blockSelector, buttonUiAtom } from 'recoils/atoms'

const TwContainer = tw.div`mb-8`

type ButtonUiBlockProps = {
  handlePageDimension: () => Promise<void>
}

export const ButtonUiBlock: React.FC<ButtonUiBlockProps> = ({
  handlePageDimension,
}) => {
  const [ref, dimension, handleDimension] = useElementDimension()

  useDimensionChange(blockSelector, dimension)

  const { dimension: btnDms } = useRecoilValue(buttonUiAtom)

  useEffect(() => {
    handleDimension()
    handlePageDimension()
  }, [
    btnDms.variant,
    btnDms.color,
    btnDms.size,
    btnDms.shape,
    btnDms.disabled,
    btnDms.icon,
    btnDms.withIcon,
    btnDms.loading,
  ])

  return (
    <TwContainer id='block' ref={ref as LegacyRef<HTMLDivElement>}>
      <Title
        title='Block'
        description={
          <>
            Button can stretch to fill the width by add <PreCode text='block' />{' '}
            prop.
          </>
        }
      />
      <ComponentDisplay
        callback={handleDimension}
        components={
          <>
            <Button label='Non Block Button' />
            <Button label='Blocked Button' block />
            <Button label='Blocked Button' block disabled />
            <Button label='Blocked Button' block loading='Loading...' />
          </>
        }
        code={`import React from "react"

export const Component = () => {
  return (
    <>
    <Button label='Non Block Button' />
    <Button label='Blocked Button' block />
    <Button label='Blocked Button' block disabled />
    <Button label='Blocked Button' block loading='Loading...' />
    </>
  )
}`}
      />
    </TwContainer>
  )
}
