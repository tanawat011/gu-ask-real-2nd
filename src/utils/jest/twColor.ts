import { hex2Rgb } from 'utils/color'

export const twColor = (hexColor: string, type?: 'bg' | 'border' | 'text') => {
  if (hexColor === 'transparent') {
    return 'transparent'
  }

  if (hexColor.length > 7) {
    return 'undefined'
  }

  const rgb = hex2Rgb(hexColor, { isArray: true }) as string[]

  if (!type) {
    type = 'bg'
  }

  return `rgb(${rgb.join(' ')} / var(--tw-${type}-opacity))`
}
