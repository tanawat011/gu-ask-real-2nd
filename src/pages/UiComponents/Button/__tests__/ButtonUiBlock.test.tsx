import { renderWithProviders } from 'utils/jest'

import { ButtonUiBlock } from '../ButtonUiBlock'

jest.mock('@fortawesome/react-fontawesome')
jest.mock('components/PageUiComponent', () => ({
  ComponentDisplay: () => <div>ComponentDisplay</div>,
  Title: () => <div>Title</div>,
}))
jest.mock('components/Button', () => ({
  Button: () => <button>Button</button>,
}))

describe('<ButtonUiBlock />', () => {
  test('renders correctly', () => {
    expect.assertions(1)

    const { asFragment } = renderWithProviders(<ButtonUiBlock />)

    expect(asFragment()).toMatchSnapshot()
  })
})
