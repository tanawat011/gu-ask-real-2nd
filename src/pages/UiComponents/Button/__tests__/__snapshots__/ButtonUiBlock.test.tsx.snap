// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonUiBlock /> renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  margin-bottom: 2rem;
}

<div
    class="c0"
  >
    <div>
      Title
    </div>
    <div>
      ComponentDisplay
    </div>
  </div>
</DocumentFragment>
`;
